File: hello-world-rz/src/ffi_slice.zig
pub fn FFISlice(comptime T: type) type {
    return extern struct {
        ptr: [*]const T,
        len: usize,
        
        const Self = @This();
        
        pub fn toSlice(self: Self) []const T {
            return self.ptr[0..self.len];
        }
    };
}

File: hello-world-rz/src/ffi_slice.rs
#[repr(C)]
pub struct FFISlice<T> {
    ptr: *const T,
    len: usize,
}

impl<'a, T> From<&'a [T]> for FFISlice<T> {
    fn from(slice: &'a [T]) -> Self
    where
        Self: 'a,
    {
        FFISlice {
            ptr: slice.as_ptr(),
            len: slice.len(),
        }
    }
}

impl<'a> From<&'a str> for FFISlice<u8> {
    fn from(slice: &'a str) -> Self
    where
        Self: 'a,
    {
        FFISlice {
            ptr: slice.as_ptr(),
            len: slice.len(),
        }
    }
}
